pkg/common/consumer/consumer_manager.go

```diff
@@ -0,0 +1,391 @@
+	groupLogger := m.logger.With(zap.String("GroupId", groupId))
+	groupLogger.Info("Creating New Managed ConsumerGroup")
+	group, err := m.factory.createConsumerGroup(groupId)
+	if err != nil {
+		groupLogger.Error("Failed To Create New Managed ConsumerGroup")
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	// Add the Sarama ConsumerGroup we obtained from the factory to the managed group map,
+	// so that it can be stopped and started via control-protocol messages.
+	m.groupLock.Lock()
+	m.groups[groupId] = createManagedGroup(group, cancel)
+	m.groupLock.Unlock()
+
+	// Begin listening on the group's Errors() channel and write them to the managedGroup's errors channel
+	m.groups[groupId].transferErrors(ctx)
+
+	// consume is passed in to the KafkaConsumerGroupFactory so that it will call the manager's
+	// consume() function instead of the one on the internal sarama ConsumerGroup.  This allows the
+	// manager to continue to block in the Consume call while a group goes through a stop/start cycle.
+	consume := func(ctx context.Context, topics []string, handler sarama.ConsumerGroupHandler) error {
+		logger.Debug("Consuming Messages On managed Consumer Group", zap.String("GroupId", groupId))
+		return m.consume(ctx, groupId, topics, handler)
+	}
+
+	// The only thing we really want from the factory is the cancel function for the customConsumerGroup
+	customGroup := m.factory.startExistingConsumerGroup(group, consume, topics, logger, handler, options...)
+	m.groups[groupId].cancelConsume = customGroup.cancel
+	return nil
+}
+
+// CloseConsumerGroup calls the Close function on the ConsumerGroup embedded in the managedGroup
+// associated with the given groupId, and also closes its managed errors channel.  It then removes the
+// group from management.
+func (m *kafkaConsumerGroupManagerImpl) CloseConsumerGroup(groupId string) error {
+	groupLogger := m.logger.With(zap.String("GroupId", groupId))
+	groupLogger.Info("Closing ConsumerGroup and removing from management")
+	managedGrp, ok := m.groups[groupId]
+	if !ok {
+		groupLogger.Warn("CloseConsumerGroup called on unmanaged group")
+		return fmt.Errorf("could not close consumer group with id '%s' - group is not present in the managed map", groupId)
+	}
+	// Make sure a managed group is "started" before closing the inner ConsumerGroup; otherwise anything
+	// waiting for the manager to restart the group will never return.
+	managedGrp.closeRestartChannel()
+	if managedGrp.cancelErrors != nil {
+		managedGrp.cancelErrors() // This will terminate any transferError call that is waiting for a restart
+	}
+	if managedGrp.cancelConsume != nil {
+		managedGrp.cancelConsume() // This will stop the factory's consume loop after the ConsumerGroup is closed
+	}
+	err := managedGrp.group.Close()
```

Comment:
should we log warnings if the `cancelErrors()` and `cancelConsume()` are nil (unexpected right?)

Reply:
Good idea; I don't think they can be non-nil unless someone creates a managedGroup without using the createManagedGroup() function.  (And guess what I did that prompted these nil-checks...)